### PR TITLE
Add C++20 support for Deathmatch project

### PR DESCRIPTION
**Pull request by TracerDS**

### CODE REVIEW

**Comment by G-Moris** [comment]
Doubtful, because of linux

**Comment by TracerDS** [comment]
> Doubtful, because of linux

it compiled fine on linux, what are you on about?

**Comment by G-Moris** [comment]
> > Doubtful, because of linux
> 
> it compiled fine on linux, what are you on about?

It is possible that problems may arise in other linuxs, it may have already been solved, it is possible that this is so

**Comment by TracerDS** [comment]
> > > Doubtful, because of linux
> > 
> > 
> > it compiled fine on linux, what are you on about?
> 
> It is possible that problems may arise in other linuxs, it may have already been solved, it is possible that this is so

If it compiles fine on x86_64-gcc-10 and arm-gcc-10 then it should compile for other modern compilers

**Comment by TheNormalnij** [comment]
> If it compiles fine on x86_64-gcc-10 and arm-gcc-10 then it should compile for other modern compilers

It doesn't work like this

**Comment by TracerDS** [comment]
> > If it compiles fine on x86_64-gcc-10 and arm-gcc-10 then it should compile for other modern compilers
> 
> It doesn't work like this

I know, but if it matches the standard then it should work on most compilers that conform to the C++ standard.
There arent any major runtime changes (beside a fix). It should not have any implications

**Comment by TheNormalnij** [comment]
> I know, but if it matches the standard then it should work on most compilers that conform to the C++ standard.
> There arent any major runtime changes (beside a fix). It should not have any implications

All compilers have different standard realization + specific bugs

**Comment by TracerDS** [comment]
> > I know, but if it matches the standard then it should work on most compilers that conform to the C++ standard.
> > There arent any major runtime changes (beside a fix). It should not have any implications
> 
> All compilers have different standard realization + specific bugs

As for this, none of the supported compilers (github actions) present a bug so we should be fine.
The only thing that could potentially be troublesome is the `[[maybe_unused]]` but even that is supported by most clang and gcc compilers.
Nothing to worry about.

I'd also assume it was thoroughly tested before rolling out C++23 version 😛

**Comment by TheNormalnij** [comment]
> As for this, none of the supported compilers (github actions) present a bug so we should be fine.
> The only thing that could potentially be troublesome is the `[[maybe_unused]]` but even that is supported by most clang and gcc compilers.
> Nothing to worry about.

You forgot about real-time bugs

**Comment by TracerDS** [comment]
> You forgot about real-time bugs

As I said before, the only real implication would be the change from `__attribute__((unused))` to `[[maybe_unused]]`.
Every other change is just an explicit SString call

File: Shared/sdk/SString.h [hunk 1]
```diff
@@ -30,6 +30,7 @@ class SString : public std::string
 
     SString(const char* szText) : std::string(szText ? szText : "") {}
 
+    template <int=0>
```
Comment (tederis) [inline Shared/sdk/SString.h:35]: What is the purpose of it?

File: Server/mods/deathmatch/logic/CPerfStat.PlayerPacketUsage.cpp [hunk 2]
```diff
@@ -408,11 +408,11 @@ void CPerfStatPlayerPacketUsageImpl::OutputTimeSpanBlock(CPerfStatResult* pResul
             const CTimeSpan* p = timeSpanList[i];
             const CTopValue& value = (p->prev.data[0].uiPktsPerSec > p->acc.data[0].uiPktsPerSec) ? p->prev.data[iTopPos] : p->acc.data[iTopPos];
 
-            row[c++] = value.uiPktsPerSec ? SString("%d", value.uiPktsPerSec) : "-";
+            row[c++] = value.uiPktsPerSec ? std::to_string(value.uiPktsPerSec) : "-";
             if (bDetail)
             {
-                row[c++] = value.uiBytesInPerSec ? SString("%d", value.uiBytesInPerSec) : "-";
-                row[c++] = value.uiBytesOutPerSec ? SString("%d", value.uiBytesOutPerSec) : "-";
+                row[c++] = value.uiBytesInPerSec ? std::to_string(value.uiBytesInPerSec) : "-";
```
Comment (tederis) [inline Server/mods/deathmatch/logic/CPerfStat.PlayerPacketUsage.cpp:414]: I see you want to use a feature from C++11 and that's great. But in this case we have an unwanted string copying. But you can combine C++11 with SString by adding a move constructor to it. The same for above and below.

Comment (TracerDS) [reply Server/mods/deathmatch/logic/CPerfStat.PlayerPacketUsage.cpp:414]: So `SString(std::to_string(...))`?

File: Shared/sdk/SString.h (hunk 1 above)
Comment (TracerDS) [reply Shared/sdk/SString.h:35]: This template hack will force C++ to properly recognize varargs and string constructors

File: Server/mods/deathmatch/logic/CPerfStat.PlayerPacketUsage.cpp (hunk 2 above)
Comment (tederis) [reply Server/mods/deathmatch/logic/CPerfStat.PlayerPacketUsage.cpp:414]: In theory, you can leave it as is. But you need to modify `SString` and add a move constructor. Like 
`SString(std::string&& strText) : std::string(std::move(strText)) {}`

File: Shared/sdk/SString.h (hunk 1 above)
Comment (tederis) [reply Shared/sdk/SString.h:35]: Please, add a comment describing it.

**Comment by tederis** [comment]
This PR is highly desirable. At least because another one PR (#3287) depends on it. I think C++20 in this module will be fine and does not cause any problems. I already have an experience with C++20 in MTA and everything is going well.

P.S. In theory there can be ABI inconsistency if some of STL structures get returned from Deathmatch module. But these potential points of inconsistency are easy to detect in most cases. At least, it can be certain that MSVC is stable enough.

**Comment by G-Moris** [comment]
Hello! Is there a problem with the support of other C++20 server modules?

**Comment by TracerDS** [comment]
> Hello! Is there a problem with the support of other C++20 server modules?

With some, probably. But I didnt check. Lets just get C++20 support for the deathmatch first

File: Server/mods/deathmatch/logic/CConsoleCommands.cpp [hunk 3]
```diff
@@ -1657,8 +1657,8 @@ bool CConsoleCommands::AuthorizeSerial(CConsole* pConsole, const char* szArgumen
 
     std::vector<SString> parts;
     SStringX(szArguments).Split(" ", parts);
-    const SString& strAccountName = parts.size() > 0 ? parts[0] : "";
-    const SString& strAction = parts.size() > 1 ? parts[1] : "";
+    const SString& strAccountName = parts.size() > 0 ? parts[0] : SString{};
+    const SString& strAction = parts.size() > 1 ? parts[1] : SString{};
```
Comment (Pirulax) [inline Server/mods/deathmatch/logic/CConsoleCommands.cpp:1661]: Really should either use `string_view` or `SString*` here instead.

File: vendor/sparsehash/src/sparsehash/internal/hashtable-common.h [hunk 4]
```diff
@@ -51,7 +51,7 @@ _START_GOOGLE_NAMESPACE_
 
 template <bool> struct SparsehashCompileAssert { };
 #define SPARSEHASH_COMPILE_ASSERT(expr, msg) \
-  __attribute__((unused)) typedef SparsehashCompileAssert<(bool(expr))> msg[bool(expr) ? 1 : -1]
+  [[maybe_unused]] typedef SparsehashCompileAssert<(bool(expr))> msg[bool(expr) ? 1 : -1]
```
Comment (Pirulax) [inline vendor/sparsehash/src/sparsehash/internal/hashtable-common.h:54]: not sure if editing vendor is a good idea

Comment (TracerDS) [reply vendor/sparsehash/src/sparsehash/internal/hashtable-common.h:54]: In this case its fine

File: Server/mods/deathmatch/logic/CConsoleCommands.cpp (hunk 3 above)
Comment (TracerDS) [reply Server/mods/deathmatch/logic/CConsoleCommands.cpp:1661]: i.e. `const std::string_view/SString* strXYZ = ... : ""/nullptr`?

**Comment by TracerDS** [comment]
Im not interested in advancing this PR further anymore.
If someone wants to refresh it, feel free to do so.

**Pull request closed without merging**